README.md
[![Build Status](https://travis-ci.org/adiwg/mdCodes.svg)](https://travis-ci.org/adiwg/mdCodes)
[![Gem Version](https://badge.fury.io/rb/adiwg-mdcodes.svg)](http://badge.fury.io/rb/adiwg-mdcodes)
[![Bower version](https://badge.fury.io/bo/mdcodes.svg)](https://badge.fury.io/bo/mdcodes)
[![npm version](https://badge.fury.io/js/mdcodes.svg)](https://badge.fury.io/js/mdcodes)

# mdCodes

mdCodes provides code lists in hash or JSON formats for loading of mdEditor and other metadata
tools that use the ADIwg [mdJson-schemas](https://github.com/adiwg/mdJson-schemas).  The code lists
include all ISO 19115-2 and ISO 19115-1 codes plus supplemental codes added by NGDC and ADIwg. The
code lists may be accessed in Ruby using the gem. Alternatively, JSON may be generated from the YAML
using [Grunt](http://gruntjs.com/getting-started). Bower and NPM packages are available as well.
See the following for instructions.

## Bower

    bower install mdcodes

## npm

    npm install mdcodes

**Note:** The npm install will also create JavaScript modules for the codelists.

## Ruby

The adiwg-mdcodes gem provides code lists in Ruby hash or JSON formats.

### Installation

Add this line to your application's Gemfile:

    gem 'adiwg-mdcodes'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install adiwg-mdcodes


### Methods

#### getYamlPath
> returns the path to the 'resources' folder containing codelist files in YAML format'.

#### getAllCodelistsDetail( returnFormat )
> returns all code lists with all codelist detail
> returnFormat = \[__hash__ | json] (string)

#### getCodelistDetail( codeListName, returnFormat )
> returns a single codelist with all code list detail
> codeListName = name of code list to return (string)
> returnFormat = \[ __hash__ | json ] (string)

#### getAllStaticCodelists( returnFormat )
> returns all static codelists with only the codelist item names
> returnFormat = \[__hash__ | json] (string)

#### getStaticCodelist( codeListName, returnFormat )
> returns a single static codelist with only the codelist item names
> codeListName = name of code list to return (string)
> returnFormat = \[__hash__ | json] (string)

##Grunt

Generate JSON verions of the code lists using Grunt.

**Note**: These instructions assume that [npm](https://docs.npmjs.com/) and [Grunt](http://gruntjs.com/getting-started) are already installed. The generated
files are not tracked by git.

1. Change to the project's root directory.
2. Install project dependencies with ```npm install```.
3. Generate the JSON code lists with ```grunt```.
4. The JSON is written to the *resources/json* directory.

## Contributing

1. Fork it (https://github.com/adiwg/mdCodes/fork)
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request


CHANGELOG.md
# Change Log

## [v2.8.0](https://github.com/adiwg/mdCodes/tree/v2.8.0)

[Full Changelog](https://github.com/adiwg/mdCodes/compare/v2.7.5...v2.8.0)

**Closed issues:**

- Add applicationProfile [\#29](https://github.com/adiwg/mdCodes/issues/29)
- Role "use" should be "user" [\#28](https://github.com/adiwg/mdCodes/issues/28)

## [v2.7.5](https://github.com/adiwg/mdCodes/tree/v2.7.5) (2019-05-16)
[Full Changelog](https://github.com/adiwg/mdCodes/compare/v2.7.4...v2.7.5)

## [v2.7.4](https://github.com/adiwg/mdCodes/tree/v2.7.4) (2019-05-16)
[Full Changelog](https://github.com/adiwg/mdCodes/compare/v2.7.2...v2.7.4)

## [v2.7.2](https://github.com/adiwg/mdCodes/tree/v2.7.2) (2018-10-17)
[Full Changelog](https://github.com/adiwg/mdCodes/compare/v2.7.1...v2.7.2)

## [v2.7.1](https://github.com/adiwg/mdCodes/tree/v2.7.1) (2018-10-10)
[Full Changelog](https://github.com/adiwg/mdCodes/compare/v2.7.0...v2.7.1)

## [v2.7.0](https://github.com/adiwg/mdCodes/tree/v2.7.0) (2018-10-03)
[Full Changelog](https://github.com/adiwg/mdCodes/compare/v2.6.3...v2.7.0)

**Closed issues:**

- Add "adiwg" code namespaces [\#27](https://github.com/adiwg/mdCodes/issues/27)
- Remove "local" from mapProjection [\#26](https://github.com/adiwg/mdCodes/issues/26)

## [v2.6.3](https://github.com/adiwg/mdCodes/tree/v2.6.3) (2018-04-16)
[Full Changelog](https://github.com/adiwg/mdCodes/compare/v2.6.2...v2.6.3)

## [v2.6.2](https://github.com/adiwg/mdCodes/tree/v2.6.2) (2018-04-05)
[Full Changelog](https://github.com/adiwg/mdCodes/compare/v2.6.1...v2.6.2)

**Closed issues:**

- Add 'observer' and 'curator' to iso-role codelist [\#24](https://github.com/adiwg/mdCodes/issues/24)

## [v2.6.1](https://github.com/adiwg/mdCodes/tree/v2.6.1) (2018-04-02)
[Full Changelog](https://github.com/adiwg/mdCodes/compare/v2.6.0...v2.6.1)

## [v2.6.0](https://github.com/adiwg/mdCodes/tree/v2.6.0) (2018-04-02)
[Full Changelog](https://github.com/adiwg/mdCodes/compare/v2.5.0...v2.6.0)

## [v2.5.0](https://github.com/adiwg/mdCodes/tree/v2.5.0) (2018-04-02)
[Full Changelog](https://github.com/adiwg/mdCodes/compare/v2.4.0...v2.5.0)

## [v2.4.0](https://github.com/adiwg/mdCodes/tree/v2.4.0) (2018-01-19)
[Full Changelog](https://github.com/adiwg/mdCodes/compare/v2.3.2...v2.4.0)

## [v2.3.2](https://github.com/adiwg/mdCodes/tree/v2.3.2) (2017-10-11)
[Full Changelog](https://github.com/adiwg/mdCodes/compare/v2.3.0...v2.3.2)

## [v2.3.0](https://github.com/adiwg/mdCodes/tree/v2.3.0) (2017-08-14)
[Full Changelog](https://github.com/adiwg/mdCodes/compare/v2.2.2...v2.3.0)

**Closed issues:**

- AssociationType to/from ScienceBase itemLink [\#23](https://github.com/adiwg/mdCodes/issues/23)
- Progress code to/from ScienceBase projectStatus [\#22](https://github.com/adiwg/mdCodes/issues/22)
- DateType to/from ScienceBase dateType [\#21](https://github.com/adiwg/mdCodes/issues/21)
- Scope code to/from ScienceBase browseCategory [\#20](https://github.com/adiwg/mdCodes/issues/20)
- OnlineFunction code to/from ScienceBase webLinkType [\#19](https://github.com/adiwg/mdCodes/issues/19)
- Role codes to/from ScienceBase contactType [\#18](https://github.com/adiwg/mdCodes/issues/18)
- Clarify associationType codes and definitions [\#17](https://github.com/adiwg/mdCodes/issues/17)

## [v2.2.2](https://github.com/adiwg/mdCodes/tree/v2.2.2) (2017-07-26)
[Full Changelog](https://github.com/adiwg/mdCodes/compare/v2.2.1...v2.2.2)

## [v2.2.1](https://github.com/adiwg/mdCodes/tree/v2.2.1) (2017-07-26)
[Full Changelog](https://github.com/adiwg/mdCodes/compare/v2.2.0...v2.2.1)

**Closed issues:**

- Namespace codelist [\#15](https://github.com/adiwg/mdCodes/issues/15)

## [v2.2.0](https://github.com/adiwg/mdCodes/tree/v2.2.0) (2017-06-29)
[Full Changelog](https://github.com/adiwg/mdCodes/compare/v2.1.7...v2.2.0)

## [v2.1.7](https://github.com/adiwg/mdCodes/tree/v2.1.7) (2017-06-21)
[Full Changelog](https://github.com/adiwg/mdCodes/compare/v2.1.6...v2.1.7)

## [v2.1.6](https://github.com/adiwg/mdCodes/tree/v2.1.6) (2017-05-16)
[Full Changelog](https://github.com/adiwg/mdCodes/compare/v2.1.5...v2.1.6)

**Implemented enhancements:**

- Add new currency codelist [\#13](https://github.com/adiwg/mdCodes/issues/13)

**Closed issues:**

- Add new associationTypes [\#14](https://github.com/adiwg/mdCodes/issues/14)
- Extend onlineFunction [\#12](https://github.com/adiwg/mdCodes/issues/12)

## [v2.1.5](https://github.com/adiwg/mdCodes/tree/v2.1.5) (2017-03-25)
[Full Changelog](https://github.com/adiwg/mdCodes/compare/v2.1.5.pre.beta2...v2.1.5)

## [v2.1.5.pre.beta2](https://github.com/adiwg/mdCodes/tree/v2.1.5.pre.beta2) (2017-03-08)
[Full Changelog](https://github.com/adiwg/mdCodes/compare/v2.1.3.pre.beta5...v2.1.5.pre.beta2)

## [v2.1.3.pre.beta5](https://github.com/adiwg/mdCodes/tree/v2.1.3.pre.beta5) (2017-02-08)
[Full Changelog](https://github.com/adiwg/mdCodes/compare/v2.1.3.pre.beta4...v2.1.3.pre.beta5)

## [v2.1.3.pre.beta4](https://github.com/adiwg/mdCodes/tree/v2.1.3.pre.beta4) (2017-02-08)
[Full Changelog](https://github.com/adiwg/mdCodes/compare/v2.1.3.pre.beta2...v2.1.3.pre.beta4)

## [v2.1.3.pre.beta2](https://github.com/adiwg/mdCodes/tree/v2.1.3.pre.beta2) (2017-02-08)
[Full Changelog](https://github.com/adiwg/mdCodes/compare/v2.1.2.pre.beta1...v2.1.3.pre.beta2)

## [v2.1.2.pre.beta1](https://github.com/adiwg/mdCodes/tree/v2.1.2.pre.beta1) (2017-02-07)
[Full Changelog](https://github.com/adiwg/mdCodes/compare/v2.1.2.pre.beta...v2.1.2.pre.beta1)

## [v2.1.2.pre.beta](https://github.com/adiwg/mdCodes/tree/v2.1.2.pre.beta) (2017-02-07)
[Full Changelog](https://github.com/adiwg/mdCodes/compare/v2.0.0...v2.1.2.pre.beta)

## [v2.0.0](https://github.com/adiwg/mdCodes/tree/v2.0.0) (2016-10-15)
[Full Changelog](https://github.com/adiwg/mdCodes/compare/v1.3.0...v2.0.0)

## [v1.3.0](https://github.com/adiwg/mdCodes/tree/v1.3.0) (2016-10-15)
[Full Changelog](https://github.com/adiwg/mdCodes/compare/v1.2.3...v1.3.0)

**Closed issues:**


test/dev.rb

# uncomment next 2 lines to run code (not gem) ....
lib = File.expand_path('lib')
res = File.expand_path('resources')
$LOAD_PATH.unshift(lib) unless $LOAD_PATH.include?(lib)
$LOAD_PATH.unshift(res)

require 'adiwg-mdcodes'
require 'json'
require 'pp'

codelists = ADIWG::Mdcodes.getCodeName('role')
pp codelists
# puts codelists.to_json

test/tc_mdcodes.rb
=begin
* Description: Test Mdcodes methods
* Author: Josh Bradley
* Date: 2014-11-07
* License: Public Domain
=end

require 'minitest/autorun'
require File.join(File.dirname(__FILE__), '..', 'lib', 'adiwg-mdcodes.rb')

class TestMdcodes < Minitest::Test

   def test_parseYaml
      @errors = Array.new

      path = ADIWG::Mdcodes.getYamlPath
      Dir["#{File.join(path, '**/*.yml')}"].each do |fname|
         begin
            YAML.load_file(fname)
         rescue Exception
            @errors << "Could not parse YAML: #{fname}"
         end
      end
      assert(@errors.empty?, @errors.join("\n"))
   end

   def test_yamlResourceDir
      yamlDir = ADIWG::Mdcodes.getYamlPath
      assert(Dir.exist?(yamlDir), 'Did not find resource Directory.')
   end

   def test_getAllCodelistsDetail
      assert_instance_of(Hash, ADIWG::Mdcodes.getAllCodelistsDetail)
   end

   def test_getCodelistDetail
      yaml = ADIWG::Mdcodes.getAllCodelistsDetail
      refute_empty(yaml)

      yaml.keys.each do |key|
         assert_instance_of(Hash, ADIWG::Mdcodes.getCodelistDetail(key), 'Failed to load ' + key)
      end
   end

   def test_getAllStaticCodelists
      assert_instance_of(Hash, ADIWG::Mdcodes.getAllStaticCodelists)
   end

   def test_getStaticCodelist
      yaml = ADIWG::Mdcodes.getAllStaticCodelists
      refute_empty(yaml)

      yaml.keys.each do |key|
         assert_instance_of(Hash, ADIWG::Mdcodes.getStaticCodelist(key), 'Failed to load ' + key)
      end
   end

   def test_getStaticCodelist_deprecated
      yaml = ADIWG::Mdcodes.getStaticCodelist('iso_associationType')
      yamlDeprecated = ADIWG::Mdcodes.getStaticCodelist('iso_associationType', 'hash', true)

      refute_empty yaml['iso_associationType']
      refute_empty yamlDeprecated['iso_associationType']
      assert (yaml['iso_associationType'].length < yamlDeprecated['iso_associationType'].length)
   end

   def test_getAllCodelistsDetail_deprecated
      yaml = ADIWG::Mdcodes.getAllCodelistsDetail
      yamlDeprecated = ADIWG::Mdcodes.getAllCodelistsDetail('hash', true)

      refute_empty yaml['iso_associationType']['codelist']
      refute_empty yamlDeprecated['iso_associationType']['codelist']
      assert (yaml['iso_associationType']['codelist'].length < yamlDeprecated['iso_associationType']['codelist'].length)
   end

end


test/tc_version.rb
require 'minitest/autorun'
require 'json'
require File.join(File.dirname(__FILE__), '..', 'lib', 'adiwg-mdcodes.rb')

class TestVersion < Minitest::Test
  def test_version
    package = JSON.load File.new(File.join(File.dirname(__FILE__), '..', 'package.json'))
    assert_equal(ADIWG::Mdcodes::VERSION, package['version'], 'Gem version and npm versions are not equal')
  end
end


Gruntfile.js
module.exports = function(grunt) {

    // Project configuration.
    grunt.initConfig({
        // This line makes your node configurations available for use
        pkg: grunt.file.readJSON('package.json'),
        convert: {
            yml2json: {
                files: [{
                    expand: true,
                    cwd: 'resources/',
                    src: ['**/*.yml'],
                    dest: 'resources/json/',
                    ext: '.json'
                }]
            }
        },
        "concat-json": {
            mdjson: {
                src: ["iso_*.json","adiwg_*.json"],
                dest: "resources/json/mdcodes.json",
                cwd: "resources/json",
                options: {
                    space: "  "
                }
            }
        },
        yamllint: {
            all: ['resources/**/*.yaml']
        }
    });
    // Each plugin must be loaded following this pattern
    grunt.loadNpmTasks('grunt-convert');
    grunt.loadNpmTasks('grunt-concat-json');
    grunt.loadNpmTasks('grunt-yamllint');
    // Custom task(s).
    // First argument names the task for use in command line
    // Second argument is a list of tasks to be run
    grunt.registerTask('default', ['yamllint:all', 'convert:yml2json', 'concat-json:mdjson']);
    grunt.registerTask('concat', ['concat-json:mdjson']);
    grunt.registerTask('lint', ['yamllint:all']);

};


lib/adiwg/mdcodes.rb
# Mdcodes - ADIwg codelists to be used with adiwgJson and mdEditor
# ... codelists are maintained in a YAML file 'mdCodes.yml'
# ... the Mdcodes module has methods to access and return codelists

# History:
#  Stan Smith 2017-08-08 add deprecated parameter to codelists
#  Stan Smith 2014-12-18 split codelists into individual YAML file
#  Stan Smith 2014-11-10 added README.md text
#  Stan Smith 2014-11-10 added support for JSON returns
#  Josh Bradley 2014-11-07 moved resources directory outside lib, add getYamlPath
#  Stan Smith 2014-11-07 add methods to return only codeNames
# 	Stan Smith 2014-11-07 original script

require 'yaml'
require 'json'

module ADIWG
   module Mdcodes

      # return the path to yaml files.
      def self.getYamlPath
         File.join(File.dirname(File.expand_path(__FILE__)), '..', '..', 'resources')
      end

      # return all codelists with all elements
      def self.getAllCodelistsDetail(format='hash', showDeprecated=false)
         path = getYamlPath + '/*.yml'
         hCodelists = {}
         Dir.glob(path) do |item|
            hCodelist = YAML.load_file(item)
            hCodelists[hCodelist['codelistName']] = hCodelist
         end
         unless showDeprecated
            hCodelists.each do |key, value|
               aKeepItems = []
               value['codelist'].each do |item|
                  if item.has_key?('deprecated')
                     unless item['deprecated']
                        aKeepItems << item
                     end
                  else
                     aKeepItems << item
                  end
                  value['codelist'] = aKeepItems
               end
            end
         end
         return hCodelists.to_json if format == 'json'
         return hCodelists
      end

      # return a single codelist with all elements
      def self.getCodelistDetail(codelist, format='hash', showDeprecated=false)
         file = File.join(getYamlPath, codelist + '.yml')
         if File.exist?(file)
            hCodelist = YAML.load_file(file)
            unless showDeprecated
               aKeepItems = []
               hCodelist['codelist'].each do |item|
                  if item.has_key?('deprecated')
                     unless item['deprecated']
                        aKeepItems << item
                     end
                  else
                     aKeepItems << item
                  end
                  hCodelist['codelist'] = aKeepItems
               end
            end
         else
            return nil
         end
         return hCodelist.to_json if format == 'json'
         return hCodelist
      end

      # return all codelist with only the codeName
      def self.getAllStaticCodelists(format='hash', showDeprecated=false)
         hCodelists = {}
         codelists = getAllCodelistsDetail('hash', showDeprecated)
         codelists.each do |key, value|
            aList = []
            value['codelist'].each do |item|
               aList << item['codeName']
            end
            hCodelists[key] = aList
         end
         return hCodelists.to_json if format == 'json'
         return hCodelists
      end

      # return a single codelist with only the codeName
      def self.getStaticCodelist(codelist, format='hash', showDeprecated=false)
         hCodelist = getCodelistDetail(codelist, 'hash', showDeprecated)
         unless hCodelist.nil?
            hCodeNames = {}
            aList = []
            hCodelist['codelist'].each do |item|
               aList << item['codeName']
            end
            hCodeNames[hCodelist['codelistName']] = aList
            return hCodeNames.to_json if format == 'json'
            return hCodeNames
         end
         return nil
      end

   end
end


lib/adiwg/mdcodes/version.rb
# adiwg mdCodes

# version 2 history
# 2.8.4 2023-03-06 added data quality
# 2.7.6 2019-07-09 fixed iso_role changed use to user
# 2.7.4 2019-04-12 fixed iso_telephone changed sourceName to CI_TelephoneTypeCode
# 2.7.4 2019-04-12 fixed iso_countries changed sourceName 'CountryCode'
# 2.7.3 2018-11-04 updated ruby version in travis.yml
# 2.7.2 2018-11-02 added 'methodology' to keywordType
# 2.7.2 2018-10-15 added 'other' to mapProjection
# 2.7.1 2018-10-06 added 'localPlanar' and 'localSystem' to mapProjection
# 2.6.2 2018-04-05 renamed iso_dataType to iso_entityType
# 2.5.0 2018-03-28 added adiwg dataTypes codelist (ANSI and SQL data types)
# 2.4.0 2018-01-10 added adiwg mapProjection codelist
# 2.3.3 2017-10-11 added associationType code 'isPartOf'
# 2.3.2 2017-10-05 added spatialRepresentation code 'coordinate'
# 2.3.1 2017-09-21 added role codes 'observer', 'curator'
# 2.3.0 2017-08-01 added showDeprecated parameter to mdCodes methods
# 2.3.0 2017-08-01 change to associationTypes: revised definitions, deprecated some codes
# 2.2.0 2017-06-29 added adiwg namespace codelist
# 2.1.7 2017-06-08 added scienceBase date codes to dateType
# 2.1.6 2017-05-16 added 'isoTopicCategory' to keyword type code list
# 2.1.5 2017-03-15 added adiwg_entityType codelist
# 2.1.4 2017-03-07 added phone types to telephone codes
# 2.1.4 2017-02-13 added adiwg_metadataStandard, adiwg_metadataRepository codelists
# 2.1.3 2017-02-07 added adiwg_addressType codelist
# 2.1.1 2016-12-18 change to European spelling 'polarisation'
# 2.1.0 2016-11-27 added 10 MI codelists

module ADIWG
   module Mdcodes
      VERSION = "2.8.4"
   end
end

# version 1 history
# 0.1.0 2014-11-05 first release
# 0.1.1 2014-11-06 added factSheet to scope
# 0.2.0 2014-11-07 add option to return only codeNames
#              ... moved resources directory outside of lib
# 0.2.1 2014-11-10 added support for JSON returns
# 0.3.0 2014-11-18 added administrator to role codelist
# 0.4.0 2014-12-18 split codelists into individual YAML file
# 1.0.0 2015-03-11 roll out as version 1
# 1.0.1 2015-06-11 added new codes to 'scope'
# 1.1.0 2015-07-22 added codelists for cellGeometry, dimensionNameType, imageCondition, coverageContentType
# 1.1.0 2015-07-28 added codelists for country, languages
# 1.2.0 2015-09-18 change source of characterSet from ISO to IANA
# 1.2.1 2015-09-28 bumped version number to match gem
